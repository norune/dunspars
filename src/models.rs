pub mod resource;

use crate::api;

use std::collections::HashMap;
use std::ops::Add;

use anyhow::Result;
use futures::stream::FuturesUnordered;
use futures::StreamExt;
use serde::{Deserialize, Serialize};

pub struct Pokemon {
    pub data: PokemonData,
    pub defense_chart: DefenseTypeChart,
    pub move_list: MoveList,
}

impl Pokemon {
    pub fn new(data: PokemonData, defense_chart: DefenseTypeChart, move_list: MoveList) -> Self {
        Self {
            data,
            defense_chart,
            move_list,
        }
    }
}

#[derive(Debug)]
pub struct PokemonData {
    pub name: String,
    pub primary_type: String,
    pub secondary_type: Option<String>,
    pub learn_moves: HashMap<String, (String, i64)>,
    pub group: PokemonGroup,
    pub game: String,
    pub generation: u8,
    pub stats: Stats,
    pub abilities: Vec<(String, bool)>,
    pub species: String,
}

impl PokemonData {
    pub async fn from_name(name: &str, game: &str) -> Result<Self> {
        api::get_pokemon(name, game).await
    }

    pub async fn get_moves(&self) -> Result<MoveList> {
        let moves_futures: FuturesUnordered<_> = self
            .learn_moves
            .iter()
            .map(|mv| Move::from_name(mv.0, self.generation))
            .collect();
        let moves_results: Vec<_> = moves_futures.collect().await;

        let mut moves = HashMap::new();
        for result in moves_results {
            let move_ = result?;
            moves.insert(move_.name.clone(), move_);
        }

        Ok(MoveList::new(moves))
    }

    pub async fn get_defense_chart(&self) -> Result<DefenseTypeChart> {
        let primary_type = Type::from_name(&self.primary_type, self.generation).await?;

        if let Some(secondary_type) = &self.secondary_type {
            let secondary_type = Type::from_name(secondary_type, self.generation).await?;

            Ok(primary_type.defense_chart + secondary_type.defense_chart)
        } else {
            Ok(primary_type.defense_chart)
        }
    }

    pub async fn get_evolution_steps(&self) -> Result<EvolutionStep> {
        api::get_evolution(&self.species).await
    }
}

#[derive(Debug)]
pub enum PokemonGroup {
    Mythical,
    Legendary,
    Regular,
}

#[derive(Default, Debug)]
pub struct Stats {
    pub hp: i64,
    pub attack: i64,
    pub defense: i64,
    pub special_attack: i64,
    pub special_defense: i64,
    pub speed: i64,
}

#[derive(Debug)]
pub struct Type {
    pub name: String,
    pub offense_chart: OffenseTypeChart,
    pub defense_chart: DefenseTypeChart,
    pub generation: u8,
}

impl Type {
    pub async fn from_name(name: &str, generation: u8) -> Result<Self> {
        api::get_type(name, generation).await
    }
}

pub const TYPES: [&str; 19] = [
    "normal", "fighting", "fire", "fighting", "water", "flying", "grass", "poison", "electric",
    "ground", "psychic", "rock", "ice", "bug", "dragon", "ghost", "dark", "steel", "fairy",
];

fn default_chart() -> HashMap<String, f32> {
    let mut chart = HashMap::new();

    for type_ in TYPES {
        chart.insert(type_.to_string(), 1.0f32);
    }

    chart
}

fn combine_charts(
    chart1: &HashMap<String, f32>,
    chart2: &HashMap<String, f32>,
) -> HashMap<String, f32> {
    let mut new_chart = HashMap::new();

    for (type_, multiplier) in chart1 {
        new_chart.insert(type_.clone(), *multiplier);
    }

    for (type_, multiplier) in chart2 {
        if let Some(new_multiplier) = new_chart.get(type_) {
            new_chart.insert(type_.clone(), multiplier * new_multiplier);
        } else {
            new_chart.insert(type_.clone(), *multiplier);
        }
    }

    new_chart
}

pub trait TypeChart {
    fn get_multiplier(&self, type_: &str) -> f32 {
        *self.get_chart().get(type_).unwrap()
    }

    fn get_chart(&self) -> &HashMap<String, f32>;
    fn get_type(&self) -> TypeCharts;
    fn get_label(&self) -> String;
    fn set_label(&mut self, label: &str);
}

pub enum TypeCharts {
    Offense,
    Defense,
}

pub trait NewTypeChart: Sized {
    fn new(chart: HashMap<String, f32>) -> Self {
        let default = default_chart();
        let new_chart = combine_charts(&default, &chart);
        Self::new_struct(new_chart)
    }

    fn new_struct(chart: HashMap<String, f32>) -> Self;
}

#[derive(Debug)]
pub struct OffenseTypeChart {
    chart: HashMap<String, f32>,
    label: String,
}
impl NewTypeChart for OffenseTypeChart {
    fn new_struct(chart: HashMap<String, f32>) -> Self {
        Self {
            chart,
            label: String::from(""),
        }
    }
}
impl TypeChart for OffenseTypeChart {
    fn get_chart(&self) -> &HashMap<String, f32> {
        &self.chart
    }

    fn get_type(&self) -> TypeCharts {
        TypeCharts::Offense
    }

    fn get_label(&self) -> String {
        self.label.clone()
    }

    fn set_label(&mut self, label: &str) {
        self.label = String::from(label);
    }
}

#[derive(Debug)]
pub struct DefenseTypeChart {
    chart: HashMap<String, f32>,
    label: String,
}
impl NewTypeChart for DefenseTypeChart {
    fn new_struct(chart: HashMap<String, f32>) -> Self {
        Self {
            chart,
            label: String::from(""),
        }
    }
}
impl TypeChart for DefenseTypeChart {
    fn get_chart(&self) -> &HashMap<String, f32> {
        &self.chart
    }

    fn get_type(&self) -> TypeCharts {
        TypeCharts::Defense
    }

    fn get_label(&self) -> String {
        self.label.clone()
    }

    fn set_label(&mut self, label: &str) {
        self.label = String::from(label);
    }
}
impl Add for DefenseTypeChart {
    type Output = DefenseTypeChart;
    fn add(self, rhs: Self) -> Self::Output {
        let chart = combine_charts(self.get_chart(), rhs.get_chart());
        let label = self.label + " " + &rhs.label;
        Self { chart, label }
    }
}

#[derive(Debug)]
pub struct Move {
    pub name: String,
    pub accuracy: Option<i64>,
    pub power: Option<i64>,
    pub pp: Option<i64>,
    pub damage_class: String,
    pub type_: String,
    pub effect: String,
    pub effect_chance: Option<i64>,
    pub generation: u8,
}

impl Move {
    pub async fn from_name(move_name: &str, generation: u8) -> Result<Self> {
        api::get_move(move_name, generation).await
    }
}

pub struct MoveList {
    value: HashMap<String, Move>,
}

impl MoveList {
    pub fn new(hashmap: HashMap<String, Move>) -> MoveList {
        MoveList { value: hashmap }
    }

    pub fn get_map(&self) -> &HashMap<String, Move> {
        &self.value
    }
}

#[derive(Debug)]
pub struct Ability {
    pub name: String,
    pub effect: String,
    pub short_effect: String,
    pub generation: u8,
}

impl Ability {
    pub async fn from_name(name: &str, generation: u8) -> Result<Self> {
        api::get_ability(name, generation).await
    }
}

#[derive(Debug, Serialize)]
pub struct EvolutionStep {
    pub name: String,
    pub methods: Vec<EvolutionMethod>,
    pub evolves_to: Vec<EvolutionStep>,
}

impl EvolutionStep {
    pub fn new(
        name: String,
        methods: Vec<EvolutionMethod>,
        evolves_to: Vec<EvolutionStep>,
    ) -> Self {
        Self {
            name,
            methods,
            evolves_to,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct EvolutionMethod {
    pub trigger: String,
    pub item: Option<String>,
    pub gender: Option<i64>,
    pub held_item: Option<String>,
    pub known_move: Option<String>,
    pub known_move_type: Option<String>,
    pub location: Option<String>,
    pub min_level: Option<i64>,
    pub min_happiness: Option<i64>,
    pub min_beauty: Option<i64>,
    pub min_affection: Option<i64>,
    pub needs_overworld_rain: Option<bool>,
    pub party_species: Option<String>,
    pub party_type: Option<String>,
    pub relative_physical_stats: Option<i64>,
    pub time_of_day: Option<String>,
    pub trade_species: Option<String>,
    pub turn_upside_down: Option<bool>,
}

impl EvolutionMethod {
    pub fn new(trigger: String) -> Self {
        Self {
            trigger,
            item: None,
            gender: None,
            held_item: None,
            known_move: None,
            known_move_type: None,
            location: None,
            min_level: None,
            min_happiness: None,
            min_beauty: None,
            min_affection: None,
            needs_overworld_rain: None,
            party_species: None,
            party_type: None,
            relative_physical_stats: None,
            time_of_day: None,
            trade_species: None,
            turn_upside_down: None,
        }
    }

    pub fn item(mut self, item: String) -> Self {
        self.item = Some(item);
        self
    }

    pub fn gender(mut self, gender: i64) -> Self {
        self.gender = Some(gender);
        self
    }

    pub fn held_item(mut self, held_item: String) -> Self {
        self.held_item = Some(held_item);
        self
    }

    pub fn known_move(mut self, known_move: String) -> Self {
        self.known_move = Some(known_move);
        self
    }

    pub fn known_move_type(mut self, known_move_type: String) -> Self {
        self.known_move_type = Some(known_move_type);
        self
    }

    pub fn location(mut self, location: String) -> Self {
        self.location = Some(location);
        self
    }

    pub fn min_level(mut self, min_level: i64) -> Self {
        self.min_level = Some(min_level);
        self
    }

    pub fn min_happiness(mut self, min_happiness: i64) -> Self {
        self.min_happiness = Some(min_happiness);
        self
    }

    pub fn min_beauty(mut self, min_beauty: i64) -> Self {
        self.min_beauty = Some(min_beauty);
        self
    }

    pub fn min_affection(mut self, min_affection: i64) -> Self {
        self.min_affection = Some(min_affection);
        self
    }

    pub fn needs_overworld_rain(mut self, needs_overworld_rain: bool) -> Self {
        self.needs_overworld_rain = Some(needs_overworld_rain);
        self
    }

    pub fn party_species(mut self, party_species: String) -> Self {
        self.party_species = Some(party_species);
        self
    }

    pub fn party_type(mut self, party_type: String) -> Self {
        self.party_type = Some(party_type);
        self
    }

    pub fn relative_physical_stats(mut self, relative_physical_stats: i64) -> Self {
        self.relative_physical_stats = Some(relative_physical_stats);
        self
    }

    pub fn time_of_day(mut self, time_of_day: String) -> Self {
        self.time_of_day = Some(time_of_day);
        self
    }

    pub fn trade_species(mut self, trade_species: String) -> Self {
        self.trade_species = Some(trade_species);
        self
    }

    pub fn turn_upside_down(mut self, turn_upside_down: bool) -> Self {
        self.turn_upside_down = Some(turn_upside_down);
        self
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Game {
    pub name: String,
    pub order: u8,
    pub generation: u8,
}

impl Game {
    pub fn new(name: String, order: u8, generation: u8) -> Self {
        Self {
            name,
            order,
            generation,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn combine_charts_test() {
        let mut chart1 = HashMap::new();
        chart1.insert("fire".to_string(), 2.0);
        chart1.insert("water".to_string(), 0.5);
        chart1.insert("steel".to_string(), 0.0);

        let mut chart2 = HashMap::new();
        chart2.insert("fire".to_string(), 2.0);
        chart2.insert("water".to_string(), 1.0);
        chart2.insert("ice".to_string(), 1.0);

        let combined = combine_charts(&chart1, &chart2);

        assert_eq!(combined.get("fire"), Some(&4.0));
        assert_eq!(combined.get("water"), Some(&0.5));
        assert_eq!(combined.get("steel"), Some(&0.0));
        assert_eq!(combined.get("ice"), Some(&1.0));
    }
}
