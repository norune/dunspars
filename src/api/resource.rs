use std::collections::HashMap;

use anyhow::{anyhow, bail, Result};
use futures::stream::FuturesOrdered;
use futures::StreamExt;
use regex::Regex;

use rustemon::client::RustemonClient;
use rustemon::games::version_group as rustemon_version;

use super::utils;
use crate::pokemon::Game;

pub enum ResourceResult {
    Valid,
    Invalid(Vec<String>),
}

#[allow(async_fn_in_trait)]
pub trait Resource: Sized {
    fn get_matches(&self, value: &str) -> Vec<String> {
        self.resource()
            .iter()
            .filter_map(|r| {
                let close_enough = if !r.is_empty() && !value.is_empty() {
                    let first_r = r.chars().next().unwrap();
                    let first_value = value.chars().next().unwrap();

                    // Only perform spellcheck on first character match; potentially expensive
                    first_r == first_value && strsim::levenshtein(r, value) < 4
                } else {
                    false
                };

                if r.contains(value) || close_enough {
                    Some(r.clone())
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
    }

    fn check(&self, value: &str) -> ResourceResult {
        let matches = self.get_matches(value);
        if matches.iter().any(|m| *m == value) {
            ResourceResult::Valid
        } else {
            ResourceResult::Invalid(matches)
        }
    }

    fn validate(&self, value: &str) -> Result<String> {
        let value = value.to_lowercase();
        match self.check(&value) {
            ResourceResult::Valid => Ok(value),
            ResourceResult::Invalid(matches) => bail!(Self::invalid_message(&value, &matches)),
        }
    }

    fn invalid_message(value: &str, matches: &[String]) -> String {
        let resource_name = Self::label();
        let mut message = format!("{resource_name} '{value}' not found.");

        if matches.len() > 20 {
            message += " Potential matches found; too many to display.";
        } else if !matches.is_empty() {
            message += &format!(" Potential matches: {}.", matches.join(" "));
        }

        message
    }

    async fn try_new(api: &RustemonClient) -> Result<Self>;
    fn resource(&self) -> Vec<String>;
    fn label() -> &'static str;
}

#[derive(Debug)]
pub struct PokemonResource {
    resource: Vec<String>,
}
impl Resource for PokemonResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_pokemon(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Pokémon"
    }
}

#[derive(Debug)]
pub struct TypeResource {
    resource: Vec<String>,
}
impl Resource for TypeResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_types(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Type"
    }
}

#[derive(Debug)]
pub struct MoveResource {
    resource: Vec<String>,
}
impl Resource for MoveResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_moves(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Move"
    }
}

#[derive(Debug)]
pub struct AbilityResource {
    resource: Vec<String>,
}
impl Resource for AbilityResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_abilities(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Ability"
    }
}

#[derive(Debug)]
pub struct GameResource {
    resource: HashMap<String, Game>,
    gen_url_regex: Regex,
}
impl Resource for GameResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        // PokéAPI keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();

        let mut resource = HashMap::new();
        let game_names = utils::get_all_games(client).await?;
        let game_data_futures: FuturesOrdered<_> = game_names
            .iter()
            .map(|g| rustemon_version::get_by_name(g, client))
            .collect();
        let game_data: Vec<_> = game_data_futures.collect().await;

        for (i, game) in game_data.into_iter().enumerate() {
            let game = game?;
            let generation = capture_gen_url(&game.generation.url, &gen_url_regex).unwrap();
            resource.insert(game.name.clone(), Game::new(game.name, i as u8, generation));
        }

        Ok(Self {
            resource,
            gen_url_regex,
        })
    }

    fn resource(&self) -> Vec<String> {
        let mut games = self.resource.iter().map(|r| r.1).collect::<Vec<&Game>>();
        games.sort_by_key(|g| g.order);

        games
            .iter()
            .map(|g| g.name.clone())
            .collect::<Vec<String>>()
    }

    fn label() -> &'static str {
        "Game"
    }
}
impl GameResource {
    pub fn get_gen(&self, game: &str) -> u8 {
        self.resource.get(game).unwrap().generation
    }

    pub fn get_gen_from_url(&self, url: &str) -> u8 {
        capture_gen_url(url, &self.gen_url_regex).unwrap()
    }
}

fn capture_gen_url(url: &str, gen_url_regex: &Regex) -> Result<u8> {
    if let Some(caps) = gen_url_regex.captures(url) {
        Ok(caps["gen"].parse::<u8>()?)
    } else {
        Err(anyhow!("Generation not found in resource url"))
    }
}
