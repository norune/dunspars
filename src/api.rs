use std::collections::HashMap;

use anyhow::{bail, Result};
use regex::Regex;

use dirs;

use rustemon::client::{CACacheManager, RustemonClient, RustemonClientBuilder};
use rustemon::games::version_group as rustemon_version;
use rustemon::moves::move_ as rustemon_moves;
use rustemon::pokemon::pokemon_species as rustemon_species;
use rustemon::pokemon::{
    ability as rustemon_ability, pokemon as rustemon_pokemon, type_ as rustemon_type,
};
use rustemon::Follow;

use rustemon::model::evolution::{
    ChainLink as RustemonEvoStep, EvolutionChain as RustemonEvoRoot,
    EvolutionDetail as RustemonEvoMethod,
};
use rustemon::model::games::VersionGroup as RustemonVersion;
use rustemon::model::moves::{Move as RustemonMove, PastMoveStatValues as RustemonPastMoveStats};
use rustemon::model::pokemon::{
    Ability as RustemonAbility, AbilityEffectChange as RustemonPastAbilityEffect,
    Pokemon as RustemonPokemon, PokemonStat as RustemonStat, PokemonType as RustemonTypeSlot,
    PokemonTypePast as RustemonPastPokemonType, Type as RustemonType,
    TypeRelations as RustemonTypeRelations, TypeRelationsPast as RustemonPastTypeRelations,
};
use rustemon::model::resource::{Effect as RustemonEffect, VerboseEffect as RustemonVerboseEffect};

use crate::pokemon::{
    Ability, EvolutionMethod, EvolutionStep, Move, PokemonData, Stats, Type, TypeChart,
};

pub struct ApiWrapper {
    client: RustemonClient,
}

impl ApiWrapper {
    pub fn try_new() -> Result<Self> {
        let mut client = RustemonClientBuilder::default();
        let mut cache_dir = if let Some(home_dir) = dirs::home_dir() {
            home_dir
        } else {
            bail!("Home directory not found")
        };
        cache_dir.push(".cache/dunspars/rustemon");

        let cache_manager = CACacheManager { path: cache_dir };
        client.with_manager(cache_manager);

        Ok(Self {
            client: client.try_build()?,
        })
    }
}

impl ApiWrapper {
    pub async fn get_pokemon(&self, pokemon: &str, game: &str) -> Result<PokemonData> {
        let RustemonPokemon {
            name,
            types,
            past_types,
            moves,
            stats,
            abilities,
            species,
            ..
        } = rustemon_pokemon::get_by_name(pokemon, &self.client).await?;
        let generation = self.get_generation(game).await?;

        let pokemon_types = self
            .match_past(generation, past_types)
            .await
            .unwrap_or(types);
        let primary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 1)
            .unwrap()
            .type_
            .name
            .clone();
        let secondary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 2)
            .map(|t| t.type_.name.clone());

        let mut learn_moves = HashMap::new();
        moves.iter().for_each(|mv| {
            let version_group = mv
                .version_group_details
                .iter()
                .find(|vg| vg.version_group.name == game);
            if let Some(vg) = version_group {
                learn_moves.insert(
                    mv.move_.name.clone(),
                    (vg.move_learn_method.name.clone(), vg.level_learned_at),
                );
            }
        });

        let abilities = abilities
            .iter()
            .map(|a| (a.ability.name.clone(), a.is_hidden))
            .collect::<Vec<_>>();

        Ok(PokemonData {
            name,
            primary_type,
            secondary_type,
            learn_moves,
            abilities,
            species: species.name,
            stats: extract_stats(stats),
            game: game.to_string(),
            generation,
            api: self,
        })
    }

    pub async fn get_type(&self, type_str: &str, generation: u8) -> Result<Type> {
        let RustemonType {
            name,
            damage_relations,
            past_damage_relations,
            ..
        } = rustemon_type::get_by_name(type_str, &self.client).await?;

        let relations = self
            .match_past(generation, past_damage_relations)
            .await
            .unwrap_or(damage_relations);

        let mut offense_chart = HashMap::new();
        let mut defense_chart = HashMap::new();

        relations.no_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 2.0);
        });

        relations.no_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 2.0);
        });

        Ok(Type {
            name,
            offense_chart: TypeChart::new(offense_chart),
            defense_chart: TypeChart::new(defense_chart),
            generation,
            api: self,
        })
    }

    pub async fn get_move(&self, name: &str, generation: u8) -> Result<Move> {
        let RustemonMove {
            name,
            mut accuracy,
            mut power,
            mut pp,
            damage_class,
            mut type_,
            mut effect_chance,
            effect_entries,
            effect_changes,
            past_values,
            ..
        } = rustemon_moves::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            mut short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_stats) = self.match_past(generation, past_values).await {
            accuracy = past_stats.accuracy.or(accuracy);
            power = past_stats.power.or(power);
            pp = past_stats.pp.or(pp);
            effect_chance = past_stats.effect_chance.or(effect_chance);

            if let Some(t) = past_stats.type_ {
                type_ = t;
            }

            if let Some(entry) = past_stats
                .effect_entries
                .into_iter()
                .find(|e| e.language.name == "en")
            {
                effect = entry.effect;
                short_effect = entry.short_effect;
            }
        }

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Move {
            name,
            accuracy,
            power,
            pp,
            damage_class: damage_class.name,
            type_: type_.name,
            effect_chance,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_ability(&self, name: &str, generation: u8) -> Result<Ability> {
        let RustemonAbility {
            name,
            effect_entries,
            effect_changes,
            ..
        } = rustemon_ability::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Ability {
            name,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_generation(&self, game: &str) -> Result<u8> {
        let RustemonVersion { generation, .. } =
            rustemon_version::get_by_name(game, &self.client).await?;

        extract_gen_from_url(&generation.url)
    }

    pub async fn get_evolution_steps(&self, species: &str) -> Result<EvolutionStep> {
        let RustemonEvoRoot { chain, .. } = rustemon_species::get_by_name(species, &self.client)
            .await?
            .evolution_chain
            .unwrap()
            .follow(&self.client)
            .await?;
        let evolution_step = traverse_chain(chain);

        Ok(evolution_step)
    }

    pub async fn get_all_moves(&self) -> Result<Vec<String>> {
        Ok(rustemon_moves::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_abilities(&self) -> Result<Vec<String>> {
        Ok(rustemon_ability::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_games(&self) -> Result<Vec<String>> {
        Ok(rustemon_version::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_types(&self) -> Result<Vec<String>> {
        Ok(rustemon_type::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_pokemon(&self) -> Result<Vec<String>> {
        Ok(rustemon_pokemon::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    async fn match_past<T: Past<U>, U>(&self, current_generation: u8, pasts: Vec<T>) -> Option<U> {
        let mut oldest_value = None;
        let mut oldest_generation = 255;

        for past in pasts {
            let past_generation = past.generation(self).await;
            if current_generation <= past_generation && past_generation <= oldest_generation {
                oldest_value = Some(past.value());
                oldest_generation = past_generation;
            }
        }

        oldest_value
    }
}

fn traverse_chain(chain_link: RustemonEvoStep) -> EvolutionStep {
    let evolution_methods = chain_link
        .evolution_details
        .into_iter()
        .map(convert_to_evolution_method)
        .collect();

    if !chain_link.evolves_to.is_empty() {
        let evolves_to = chain_link
            .evolves_to
            .into_iter()
            .map(traverse_chain)
            .collect();

        EvolutionStep::new(chain_link.species.name, evolution_methods, evolves_to)
    } else {
        EvolutionStep::new(chain_link.species.name, evolution_methods, vec![])
    }
}

fn convert_to_evolution_method(evolution: RustemonEvoMethod) -> EvolutionMethod {
    let mut method = EvolutionMethod::new(evolution.trigger.name);
    if let Some(item) = evolution.item {
        method = method.item(item.name);
    }
    if let Some(gender) = evolution.gender {
        method = method.gender(gender);
    }
    if let Some(held_item) = evolution.held_item {
        method = method.held_item(held_item.name);
    }
    if let Some(known_move) = evolution.known_move {
        method = method.known_move(known_move.name);
    }
    if let Some(known_move_type) = evolution.known_move_type {
        method = method.known_move_type(known_move_type.name);
    }
    if let Some(location) = evolution.location {
        method = method.location(location.name);
    }
    if let Some(min_level) = evolution.min_level {
        method = method.min_level(min_level);
    }
    if let Some(min_happiness) = evolution.min_happiness {
        method = method.min_happiness(min_happiness);
    }
    if let Some(min_beauty) = evolution.min_beauty {
        method = method.min_beauty(min_beauty);
    }
    if let Some(min_affection) = evolution.min_affection {
        method = method.min_affection(min_affection);
    }
    if let Some(party_species) = evolution.party_species {
        method = method.party_species(party_species.name);
    }
    if let Some(party_type) = evolution.party_type {
        method = method.party_type(party_type.name);
    }
    if let Some(relative_physical_stats) = evolution.relative_physical_stats {
        method = method.relative_physical_stats(relative_physical_stats);
    }
    if let Some(trade_species) = evolution.trade_species {
        method = method.trade_species(trade_species.name);
    }
    if evolution.needs_overworld_rain {
        method = method.needs_overworld_rain(true);
    }
    if evolution.turn_upside_down {
        method = method.turn_upside_down(true);
    }
    if !evolution.time_of_day.is_empty() {
        method = method.time_of_day(evolution.time_of_day);
    }

    method
}

fn extract_gen_from_url(url: &str) -> Result<u8> {
    let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();
    let caps = gen_url_regex.captures(url).unwrap();
    let generation = caps["gen"].parse::<u8>()?;
    Ok(generation)
}

fn extract_stats(stats_vec: Vec<RustemonStat>) -> Stats {
    let mut stats = Stats::default();

    for RustemonStat {
        stat, base_stat, ..
    } in stats_vec
    {
        match stat.name.as_str() {
            "hp" => stats.hp = base_stat,
            "attack" => stats.attack = base_stat,
            "defense" => stats.defense = base_stat,
            "special-attack" => stats.special_attack = base_stat,
            "special-defense" => stats.special_defense = base_stat,
            "speed" => stats.speed = base_stat,
            _ => (),
        }
    }

    stats
}

trait Past<T> {
    async fn generation(&self, api: &ApiWrapper) -> u8;
    fn value(self) -> T;
}

impl Past<Vec<RustemonTypeSlot>> for RustemonPastPokemonType {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> Vec<RustemonTypeSlot> {
        self.types
    }
}

impl Past<RustemonTypeRelations> for RustemonPastTypeRelations {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> RustemonTypeRelations {
        self.damage_relations
    }
}

impl Past<RustemonPastMoveStats> for RustemonPastMoveStats {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> RustemonPastMoveStats {
        self
    }
}

impl Past<Vec<RustemonEffect>> for RustemonPastAbilityEffect {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> Vec<RustemonEffect> {
        self.effect_entries
    }
}
