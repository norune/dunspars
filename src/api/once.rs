use crate::resource::GameResource;
use regex::Regex;
use std::sync::OnceLock;

pub fn game_resource() -> &'static GameResource {
    static GAME_RESOURCE: OnceLock<GameResource> = OnceLock::new();
    GAME_RESOURCE.get_or_init(|| GameResource::try_new().unwrap())
}

pub fn gen_url_regex() -> &'static Regex {
    static GEN_URL_REGEX: OnceLock<Regex> = OnceLock::new();
    GEN_URL_REGEX.get_or_init(|| {
        // Pok√©API keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap()
    })
}
