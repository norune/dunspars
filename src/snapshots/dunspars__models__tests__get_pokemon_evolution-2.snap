---
source: src/models.rs
expression: applin
---
name: applin
methods: []
evolves_to:
  - name: flapple
    methods:
      - trigger: use-item
        item: tart-apple
        gender: ~
        held_item: ~
        known_move: ~
        known_move_type: ~
        location: ~
        min_level: ~
        min_happiness: ~
        min_beauty: ~
        min_affection: ~
        needs_overworld_rain: ~
        party_species: ~
        party_type: ~
        relative_physical_stats: ~
        time_of_day: ~
        trade_species: ~
        turn_upside_down: ~
    evolves_to: []
  - name: appletun
    methods:
      - trigger: use-item
        item: sweet-apple
        gender: ~
        held_item: ~
        known_move: ~
        known_move_type: ~
        location: ~
        min_level: ~
        min_happiness: ~
        min_beauty: ~
        min_affection: ~
        needs_overworld_rain: ~
        party_species: ~
        party_type: ~
        relative_physical_stats: ~
        time_of_day: ~
        trade_species: ~
        turn_upside_down: ~
    evolves_to: []
  - name: dipplin
    methods: []
    evolves_to:
      - name: hydrapple
        methods: []
        evolves_to: []
